### PR TITLE
statusBar: focusNext and focusPrevious

### LINKED ISSUE #95433: Status bar items should be keyboard navigable
Once focus is on the status bar it should be possible to navigate left / right via keyboard so that focus jump over each element.

Destilled from https://github.com/microsoft/vscode/issues/95428
@bpasero let me know if you have ideas, pointers. We might run into the issue that you reverse the elements for your flex layout.

### PR DESCRIPTION
**Pull request by isidorn**
Fixes https://github.com/microsoft/vscode/issues/95433

This PR introduces:
* `statusBarFocus` context key
* `workbench.statusBar.focusPrevious` and `workbench.statusBar.focusNext` commands
* Uses the status hover background color for showing focus
* Exposes `focusNextEntry` and `focusPreviousEntry` on the `statusBarService`
* focus goes cyclic, so if you focus next and the last is focused we focus the first.
* Tab moves focus out of the status bar. 

All with the idea to make the status bar item navigable once the status bar is focused. So the flow is `F6` to focus status bar, `left` and `right` to go through the status bar items.
`Up` and `down` are secondary keybindings for `left` and `right`. MS Word behaves like that and I think it makes sense to also navigate like this to be aligned with activity bar and users might not know how status bar looks visually.

Also I decided to put focus on the `a` element, because that way Orca announces the most concise `aria-label`.

I tried this with Orca, NVDA and VoiceOver and seems to work nicely.

Two issues I hit:
* On Linux: `entries` are not ordered with how they are displayed. Though I could not repro this on Mac or Win. @bpasero might you know more (I know you use that flax trick but I thought you just use it for the display, not on the model level)
* Once an item is focused VoiceOver reads the ariaLabel of the item and the whole ariaLabel of the statusBar. This seems to work fine with Orca and NVDA. So we can file an issue with VoiceOver.

@bpasero let me know what you think and if it looks good to you I will merge. Thanks!

### CODE REVIEW

**Comment by akash07k** [comment]
Seems promising. :-) Thanks

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 1]
```diff
@@ -645,13 +694,14 @@ class StatusbarEntryItem extends Disposable {
 
 	private entry!: IStatusbarEntry;
 
-	private labelContainer!: HTMLElement;
+	public labelContainer!: HTMLElement;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:697]:
```suggestion
	labelContainer!: HTMLElement;
```

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 2]
```diff
@@ -442,8 +481,18 @@ export class StatusbarPart extends Part implements IStatusbarService {
...
+		this.viewModel.focusNextEntry();
+	}
+
+	focusPreviousEntry(): void {
+		this.viewModel.focusPreviousEntry();
+	}
+
 	createContentArea(parent: HTMLElement): HTMLElement {
 		this.element = parent;
+		const scopedContextKeyService = this.contextKeyService.createScoped(this.element);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:494]:
```suggestion
		
		// Track focus within container
		const scopedContextKeyService = this.contextKeyService.createScoped(this.element);
```

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 3]
```diff
@@ -711,11 +761,19 @@ class StatusbarEntryItem extends Disposable {
...
 		// Update: Command
 		if (!this.entry || entry.command !== this.entry.command) {
-			this.commandListener.clear();
+			this.commandMouseListener.clear();
+			this.commandKeyboardListener.clear();
 
 			const command = entry.command;
 			if (command) {
-				this.commandListener.value = addDisposableListener(this.labelContainer, EventType.CLICK, () => this.executeCommand(command));
+
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:769]:
```suggestion
```

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts [hunk 4]
```diff
@@ -51,8 +56,11 @@ interface IStatusbarViewModelEntry {
 	alignment: StatusbarAlignment;
 	priority: number;
 	container: HTMLElement;
+	labelContainer: HTMLElement;
 }
 
+const CONTEXT_STATUS_BAR_FOCUSED = new RawContextKey<boolean>('statusBarFocused', true);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:62]: The default should be `false`, no?

**Review by bpasero — CHANGES_REQUESTED** [review]
@isidorn only minor feedback, but I struggled how to try this out. Tried with pressing F6 but I am not seeing focus well enough to know where I am focused.

File: src/vs/workbench/browser/parts/statusbar/statusbarPart.ts (hunk 4 above)
Comment (isidorn) [reply src/vs/workbench/browser/parts/statusbar/statusbarPart.ts:62]: Yes, I changed this for testing and forgot to turn it back to `false`!

**Comment by isidorn** [comment]
@bpasero thanks for the review, I will adress your comments and merge. 
We can introduce a new color for focus. Though since this will primarily be used by screen reader users I would only introduce it once users request it.

The original F6 focus on the status bar is not rendered - we should probably change that.
As soon as you press left / right you should see the focus as the hover.

**Comment by bpasero** [comment]
@isidorn I see it working now, but when I press F6 I would expect the first item in the status bar to show focus, but in fact the last one seems to have focus?

**Comment by isidorn** [comment]
@bpasero the behavior is first the whole status bar is focused (so everything first gets read out to the user), then pressing left / right takes you through each item.
I do not see the last item being focus just by pressing F6.

**Comment by bpasero** [comment]
recording

Here is what I see.

**Comment by isidorn** [comment]
@bpasero that's strange. Can not see that. Can you maybe do some
`setInterval(() => console.log(document.activeElement), 1000)` and check where the focus gets lost

**Comment by bpasero** [comment]
@isidorn there you go:

image

I think the reason for this is that my first entry "Remote" is always hidden.

**Comment by isidorn** [comment]
@bpasero thanks. I have pushed a fix, that we jupm over hidden items, made the code a bit more complex but works fine now.
I have also changed the initial value of the context key.

Thus I plan to merge this in if it is ok by you.
As follow up work I plan to change the F6 navigation to also focus onto tha activity bar and the panel title area.

**Comment by isidorn** [comment]
For reference here's the commit that tackles that F6 navigates to the activity bar ef14938a46

**Review by bpasero — APPROVED** [review]
LGTM. I feel that pressing `Escape` key should also bring me out of this mode.

**Comment by isidorn** [comment]
@bpasero thanks, merging in.
Not so sure about `Esc` since we already changed `ActionBar` to not do anything on Escape - since the focus would jump unexpectedly.
Though we should keep the whole "what to do with `esc` regarding focus" discussion on-going.

**Pull request merged**